File: cpp/crc32.cc
#include <iostream>
#include <fstream>
#include <vector>
#include <iomanip>
#include <inttypes.h>

const static int ONE_MB = 1024 * 1024;

class Crc32 {
private:
  const static uint32_t seed;
  std::vector<uint32_t> table;
  uint32_t value;
public:
  Crc32();

  uint32_t crc(const std::vector<uint8_t>& bytes);
  void     start();
  void     update(const std::vector<uint8_t>& bytes);
  uint32_t finalize();
};

const uint32_t Crc32::seed = 0xEDB88320;

Crc32::Crc32() : value(0)
{
  uint32_t v;
  table.resize(256);

  for(size_t i = 0; i < 256; ++i) {
    v = i;
    for(size_t j = 0; j < 8; ++j) {
      v = (v & 1) ? (seed ^ (v >> 1)) : (v >> 1);
    }
    table[i] = v;
  }
}

uint32_t Crc32::crc(const std::vector<uint8_t>& bytes)
{
  start();
  update(bytes);
  return finalize();
}

void Crc32::start()
{
  value = 0xFFFFFFFF;
}

void Crc32::update(const std::vector<uint8_t>& bytes)
{
  for(std::vector<uint8_t>::const_iterator i = bytes.begin();
      i != bytes.end();
      i++) {
    value = table[(value ^ *i) & 0xFF] ^ (value >> 8);
  }
}

uint32_t Crc32::finalize()
{
  return value ^ 0xFFFFFFFFul;
}


int main(int argc, char* argv[])
{
  Crc32 crc;
  std::vector<uint8_t> buf;

  buf.resize(ONE_MB);


  for(int i = 1; i < argc; ++i) {
    crc.start();
    std::ifstream f(argv[i], std::ios::binary);

    while(f.read((char*)&buf[0], ONE_MB)) {
      crc.update(buf);
    }

    buf.resize(f.gcount());
    crc.update(buf);

    f.close();
    std::cout << argv[i] << ": " << std::setw(8) << std::setfill('0') <<
      std::hex << crc.finalize() << std::endl;
  }
}

File: java/Crc32.java
import java.io.FileInputStream;
import java.io.IOException;
import java.io.FileNotFoundException;

public class Crc32 {

    private int value;
    private int[] table = new int[256];

    private static final int SEED = 0xEDB88320;

    public Crc32() {

        for(int i = 0; i < 256; ++i) {
            int v = i;
            for(int j = 0; j < 8; ++j) {
                v = (v & 1) == 1 ? (SEED ^ (v >>> 1)) : (v >>> 1);
            }
            this.table[i] = v;
        }
    }

    public void start() {
        value = 0xFFFFFFFF;
    }

    public void update(byte[] bytes, int length) {
        for(int i = 0; i < length; ++i) {
            this.value = this.table[((this.value ^ bytes[i]) & 0xFF)] ^ (this.value >>> 8);
        }
    }


    public int result() {
        return (this.value ^ 0xFFFFFFFF);
    }

    public static void main(String[] args) throws IOException {
        Crc32 crc  = new Crc32();
        byte[] buf = new byte[1024 * 1024];
        int len;

        for (String filename : args) {
            FileInputStream fs = null;
            try {
                fs = new FileInputStream(filename);
                crc.start();
                while((len = fs.read(buf)) > 0) {
                    crc.update(buf, len);
                }

                System.out.format("%s: %08x\n", filename, crc.result());

            } catch (FileNotFoundException ex) {

            } finally {
                if (fs != null) {
                    fs.close();
                }
            }
        }

    }
}


